src/stats.ts

```diff
@@ -0,0 +1,187 @@
+import { Client } from './client';
+import { GenericObject, StatsEvent } from './types';
+import { isDfid } from './util/is';
+
+interface BaseStatsParams {
+  sessionId: string;
+  hashid?: string;
+}
+
+interface ClickStatsParams extends BaseStatsParams {
+  id: string;
+  datatype?: string;
+  query?: string;
+  customResultsId?: string | number;
+}
+
+interface BannerStatsParams extends BaseStatsParams {
+  bannerId: number;
+}
+
+export class StatsClient {
+  private client: Client;
+
+  public constructor(client: Client) {
+    this.client = client;
+  }
+
+  /**
+   * Allows changing the current client we
+   * are using
+   *
+   * @param  {Object}   client    The new Doofinder Client to use
+   *
+   */
+  public use(client: Client): void {
+    this.client = client;
+  }
+
+  /**
+   * Registers a new session with Doofinder, and returns
+   * the session ID generated for this session
+   *
+   * @param   {String}     sessionId   The sessionId
+   *
+   * @return  {String}     The newly generated Session ID
+   */
+  public async requestSession({ sessionId, hashid }: BaseStatsParams): Promise<GenericObject> {
+    if (!sessionId) {
+      throw new Error('Session ID must be defined');
+    }
+
+    if (hashid) {
+      this.client.hashid = hashid;
+    }
+    return await this.client.stats(StatsEvent.Init, { session_id: sessionId });
```

Comment:
You don't have to _await_ here, just return the same promise the client returns. The same for the other functions.